README.md
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

## Introduction

Epilepsia is a user friendly cape for the [beaglebone](https://beagleboard.org/black) capable of driving up to 32 strips of [neopixels](https://learn.adafruit.com/adafruit-neopixel-uberguide) (actually, WS2812B LEDs) in parallel.

![](http://guigui.us/epilepsia/images/img1.jpg)
![](http://guigui.us/epilepsia/images/img2.jpg)
![](http://guigui.us/epilepsia/images/img3.jpg)
![](http://guigui.us/epilepsia/gifs/demo2small.gif)

## Features

 * Can drive 64x32 WS2812B LEDs at around 450-500 fps
 * Support the [Open Pixel Control](http://openpixelcontrol.org/) protocol and websockets for feeding data.
 * Can be configured for 8, 16 or 32 outputs.
 * Gamma correction and brightness settings

## Architecture

The [led_driver](https://github.com/fyhertz/epilepsia/blob/master/arm/leddriver.cpp) class handles the communication with the PRUs. Each PRU can drive 16 WS2812 led strips in parallel using two 74HC4094 serial to parallel shift register. The bits of the frame buffer are reordered by the beablebone's CPU before being copied to the PRUs shared memory. Each PRU just has to sequentially read that memory to fill its shift registers. The single wire protocol of the led strips is implemented by switching the parallel outputs of the 74HC4094 IC.

The following schematic shows the wiring the 74HC4094s and the PRUs:

![Schematic](https://raw.githubusercontent.com/fyhertz/epilepsia/master/schematics/schematic.png)

## Build instructions

**You can compile epilepsia and build a debian package with the provided Dockerfile:** 
```
git clone https://github.com/fyhertz/epilepsia.git
cd epilepsia
docker build -t fyhertz/epilepsia -f docker/Dockerfile docker
docker run -ti -v `pwd -P`:/data fyhertz/epilepsia
```

The [-v option](https://docs.docker.com/storage/volumes/) is needed to mount the project files in the container. On Windows/Mac OS you might want to replace the `pwd` command with the path to the your clone of the project.

**Cross compiling epilepsia from a linux host:**

1. You will need the clpru C/C++03 compiler from TI to build the PRU firmware:
```
wget http://software-dl.ti.com/codegen/esd/cgt_public_sw/PRU/2.2.1/ti_cgt_pru_2.2.1_linux_installer_x86.bin -O clpru.bin
chmod +x clpru.bin
./clpru.bin --mode unattended --prefix /opt/ti
```

2. A C++14 cross compiler is needed for building the ARM linux binary. On debian/ubuntu run:
```
sudo apt-get update && apt-get install -y make g++-arm-linux-gnueabihf
```

3. The project simply uses some plain Makefiles
```
make
```

## Installation instructions

Supported hardware: [beaglebone black](https://beagleboard.org/black), [beaglebone black wireless](https://beagleboard.org/black-wireless), [beaglebone green](https://beagleboard.org/green), [beaglebone green wireless](https://beagleboard.org/green-wireless)

For simplicity sake, this guide assumes that you are using, or are willing to use, a recent firmware image for your device. **As of May 2018**, that means: 
 - **kernel version > 4.9.** Before that, the pru_rproc driver [behaves differently](https://groups.google.com/d/msg/beagleboard/4P9NdglojBo/qqizuGCZAQAJ)
 - **U-boot overlays** which have recently [replaced Kernel overlays](https://groups.google.com/d/msg/beagleboard/P_Y5yjJyuu4/yaZfkXfAAgAJ)
 - **[cape-universal](https://github.com/cdsteinkuehler/beaglebone-universal-io) readily available.** It's a set of device tree overlays that allows you to easily select pin modes with the pin-config command line tool.
 
 If for some reason you don't want to reflash your beaglebone, epilepsia should still work (maybe with some more work) as long as you have the pru_rproc driver (and not the uio_pruss driver).

1. So, if you don't want any headache, get the latest official debian build [here](https://beagleboard.org/latest-images) (with debian 9 as of May 2018) and flash your device with it.
2. On the bb green wireless you need to disable the wireless virtual overlay. Epilepsia requires P8_11 and P9_31. To do that edit /boot/uEnv.txt and **uncomment "disable_uboot_overlay_wireless=1"**. On the bb black, disable hdmi video: **uncomment "disable_uboot_overlay_video=1"**.
3. TODO

## Optional instructions

To avoid unnecessary write access to bb emmc and prolongate its lifespan you can do the following:

1. Use a circular buffer for syslogs: `apt-get install busybox-syslogd; dpkg --purge rsyslog`
2. Use tmpfs for logs and tmp files. Edit /etc/fstab and append:
```
tmpfs           /tmp            tmpfs   nosuid,nodev         0       0
tmpfs           /var/log        tmpfs   nosuid,nodev         0       0
tmpfs           /var/tmp        tmpfs   nosuid,nodev         0       0
```
3. At this point you can remove logrotate `apt remove --purge logrotate`
 
## TODO

 * Software estimation of power consumption
 * Beaglepocket support
 * Frame interpolation


debian/changelog
epilepsia (0.3) unstable; urgency=medium

  * Add OPC command to exit epilepsia
  * Update systemd service to shutdown beaglebone when epilepsia exits

 -- Simon <fyhertz@gmail.com>  Sun, 31 Dec 2023 00:41:01 +0100

epilepsia (0.2) unstable; urgency=medium

  * bug fix when saving settings

 -- Simon <fyhertz@gmail.com>  Sat, 07 Dec 2019 17:07:25 +0100

epilepsia (0.1) unstable; urgency=medium

  * Initial release.

 -- Simon <fyhertz@gmail.com>  Sat, 16 Mar 2019 18:35:46 +0100


arm/sha1.hpp
/*
 * Needed for the websocket protocol handshake
 * https://github.com/983/SHA1/blob/master/sha1.hpp
 */

#include <stdint.h>
#include <string.h>

#define SHA1_HEX_SIZE (40 + 1)
#define SHA1_BASE64_SIZE (28 + 1)

class sha1 {
private:
    void add_byte_dont_count_bits(uint8_t x)
    {
        buf[i++] = x;

        if (i >= sizeof(buf)) {
            i = 0;
            process_block(buf);
        }
    }

    static uint32_t rol32(uint32_t x, uint32_t n)
    {
        return (x << n) | (x >> (32 - n));
    }

    static uint32_t make_word(const uint8_t* p)
    {
        return ((uint32_t)p[0] << 3 * 8) | ((uint32_t)p[1] << 2 * 8) | ((uint32_t)p[2] << 1 * 8) | ((uint32_t)p[3] << 0 * 8);
    }

    void process_block(const uint8_t* ptr)
    {
        const uint32_t c0 = 0x5a827999;
        const uint32_t c1 = 0x6ed9eba1;
        const uint32_t c2 = 0x8f1bbcdc;
        const uint32_t c3 = 0xca62c1d6;

        uint32_t a = state[0];
        uint32_t b = state[1];
        uint32_t c = state[2];
        uint32_t d = state[3];
        uint32_t e = state[4];

        uint32_t w[16];

        for (int i = 0; i < 16; i++)
            w[i] = make_word(ptr + i * 4);

#define SHA1_LOAD(i) w[i & 15] = rol32(w[(i + 13) & 15] ^ w[(i + 8) & 15] ^ w[(i + 2) & 15] ^ w[i & 15], 1);
#define SHA1_ROUND_0(v, u, x, y, z, i)                       \
    z += ((u & (x ^ y)) ^ y) + w[i & 15] + c0 + rol32(v, 5); \
    u = rol32(u, 30);
#define SHA1_ROUND_1(v, u, x, y, z, i)                       \
    SHA1_LOAD(i)                                             \
    z += ((u & (x ^ y)) ^ y) + w[i & 15] + c0 + rol32(v, 5); \
    u = rol32(u, 30);
#define SHA1_ROUND_2(v, u, x, y, z, i)               \
    SHA1_LOAD(i)                                     \
    z += (u ^ x ^ y) + w[i & 15] + c1 + rol32(v, 5); \
    u = rol32(u, 30);
#define SHA1_ROUND_3(v, u, x, y, z, i)                             \
    SHA1_LOAD(i)                                                   \
    z += (((u | x) & y) | (u & x)) + w[i & 15] + c2 + rol32(v, 5); \
    u = rol32(u, 30);
#define SHA1_ROUND_4(v, u, x, y, z, i)               \
    SHA1_LOAD(i)                                     \
    z += (u ^ x ^ y) + w[i & 15] + c3 + rol32(v, 5); \
    u = rol32(u, 30);

        SHA1_ROUND_0(a, b, c, d, e, 0);
        SHA1_ROUND_0(e, a, b, c, d, 1);
        SHA1_ROUND_0(d, e, a, b, c, 2);
        SHA1_ROUND_0(c, d, e, a, b, 3);
        SHA1_ROUND_0(b, c, d, e, a, 4);
        SHA1_ROUND_0(a, b, c, d, e, 5);
        SHA1_ROUND_0(e, a, b, c, d, 6);
        SHA1_ROUND_0(d, e, a, b, c, 7);
        SHA1_ROUND_0(c, d, e, a, b, 8);
        SHA1_ROUND_0(b, c, d, e, a, 9);
        SHA1_ROUND_0(a, b, c, d, e, 10);
        SHA1_ROUND_0(e, a, b, c, d, 11);
        SHA1_ROUND_0(d, e, a, b, c, 12);
        SHA1_ROUND_0(c, d, e, a, b, 13);
        SHA1_ROUND_0(b, c, d, e, a, 14);
        SHA1_ROUND_0(a, b, c, d, e, 15);
        SHA1_ROUND_1(e, a, b, c, d, 16);
        SHA1_ROUND_1(d, e, a, b, c, 17);
        SHA1_ROUND_1(c, d, e, a, b, 18);
        SHA1_ROUND_1(b, c, d, e, a, 19);
        SHA1_ROUND_2(a, b, c, d, e, 20);
        SHA1_ROUND_2(e, a, b, c, d, 21);
        SHA1_ROUND_2(d, e, a, b, c, 22);
        SHA1_ROUND_2(c, d, e, a, b, 23);
        SHA1_ROUND_2(b, c, d, e, a, 24);
        SHA1_ROUND_2(a, b, c, d, e, 25);
        SHA1_ROUND_2(e, a, b, c, d, 26);
        SHA1_ROUND_2(d, e, a, b, c, 27);
        SHA1_ROUND_2(c, d, e, a, b, 28);
        SHA1_ROUND_2(b, c, d, e, a, 29);
        SHA1_ROUND_2(a, b, c, d, e, 30);
        SHA1_ROUND_2(e, a, b, c, d, 31);
        SHA1_ROUND_2(d, e, a, b, c, 32);
        SHA1_ROUND_2(c, d, e, a, b, 33);
        SHA1_ROUND_2(b, c, d, e, a, 34);
        SHA1_ROUND_2(a, b, c, d, e, 35);
        SHA1_ROUND_2(e, a, b, c, d, 36);
        SHA1_ROUND_2(d, e, a, b, c, 37);
        SHA1_ROUND_2(c, d, e, a, b, 38);
        SHA1_ROUND_2(b, c, d, e, a, 39);
        SHA1_ROUND_3(a, b, c, d, e, 40);
        SHA1_ROUND_3(e, a, b, c, d, 41);
        SHA1_ROUND_3(d, e, a, b, c, 42);
        SHA1_ROUND_3(c, d, e, a, b, 43);
        SHA1_ROUND_3(b, c, d, e, a, 44);
        SHA1_ROUND_3(a, b, c, d, e, 45);
        SHA1_ROUND_3(e, a, b, c, d, 46);
        SHA1_ROUND_3(d, e, a, b, c, 47);
        SHA1_ROUND_3(c, d, e, a, b, 48);
        SHA1_ROUND_3(b, c, d, e, a, 49);
        SHA1_ROUND_3(a, b, c, d, e, 50);
        SHA1_ROUND_3(e, a, b, c, d, 51);
        SHA1_ROUND_3(d, e, a, b, c, 52);
        SHA1_ROUND_3(c, d, e, a, b, 53);
        SHA1_ROUND_3(b, c, d, e, a, 54);
        SHA1_ROUND_3(a, b, c, d, e, 55);
        SHA1_ROUND_3(e, a, b, c, d, 56);
        SHA1_ROUND_3(d, e, a, b, c, 57);
        SHA1_ROUND_3(c, d, e, a, b, 58);
        SHA1_ROUND_3(b, c, d, e, a, 59);
        SHA1_ROUND_4(a, b, c, d, e, 60);
        SHA1_ROUND_4(e, a, b, c, d, 61);
        SHA1_ROUND_4(d, e, a, b, c, 62);
        SHA1_ROUND_4(c, d, e, a, b, 63);
        SHA1_ROUND_4(b, c, d, e, a, 64);
        SHA1_ROUND_4(a, b, c, d, e, 65);
        SHA1_ROUND_4(e, a, b, c, d, 66);
        SHA1_ROUND_4(d, e, a, b, c, 67);
        SHA1_ROUND_4(c, d, e, a, b, 68);
        SHA1_ROUND_4(b, c, d, e, a, 69);
        SHA1_ROUND_4(a, b, c, d, e, 70);
        SHA1_ROUND_4(e, a, b, c, d, 71);
        SHA1_ROUND_4(d, e, a, b, c, 72);
        SHA1_ROUND_4(c, d, e, a, b, 73);
        SHA1_ROUND_4(b, c, d, e, a, 74);
        SHA1_ROUND_4(a, b, c, d, e, 75);
        SHA1_ROUND_4(e, a, b, c, d, 76);
        SHA1_ROUND_4(d, e, a, b, c, 77);
        SHA1_ROUND_4(c, d, e, a, b, 78);
        SHA1_ROUND_4(b, c, d, e, a, 79);

#undef SHA1_LOAD
#undef SHA1_ROUND_0
#undef SHA1_ROUND_1
#undef SHA1_ROUND_2
#undef SHA1_ROUND_3
#undef SHA1_ROUND_4

        state[0] += a;
        state[1] += b;
        state[2] += c;
        state[3] += d;
        state[4] += e;
    }

public:
    uint32_t state[5];
    uint8_t buf[64];
    uint32_t i;
    uint64_t n_bits;

    sha1(const char* text = NULL)
        : i(0)
        , n_bits(0)
    {
        state[0] = 0x67452301;
        state[1] = 0xEFCDAB89;
        state[2] = 0x98BADCFE;
        state[3] = 0x10325476;
        state[4] = 0xC3D2E1F0;
        add(text);
    }

    sha1& add(uint8_t x)
    {
        add_byte_dont_count_bits(x);
        n_bits += 8;
        return *this;
    }

    sha1& add(char c)
    {
        return add(*(uint8_t*)&c);
    }

    sha1& add(const void* data, uint32_t n)
    {
        if (!data)
            return *this;

        const uint8_t* ptr = (const uint8_t*)data;

        // fill up block if not full
        for (; n && i % sizeof(buf); n--)
            add(*ptr++);

        // process full blocks
        for (; n >= sizeof(buf); n -= sizeof(buf)) {
            process_block(ptr);
            ptr += sizeof(buf);
            n_bits += sizeof(buf) * 8;
        }

        // process remaining part of block
        for (; n; n--)
            add(*ptr++);

        return *this;
    }

    sha1& add(const char* text)
    {
        return add(text, strlen(text));
    }

    sha1& finalize()
    {
        // hashed text ends with 0x80, some padding 0x00 and the length in bits
        add_byte_dont_count_bits(0x80);
        while (i % 64 != 56)
            add_byte_dont_count_bits(0x00);
        for (int j = 7; j >= 0; j--)
            add_byte_dont_count_bits(n_bits >> j * 8);

        return *this;
    }

    const sha1& print_hex(
        char* hex,
        bool zero_terminate = true,
        const char* alphabet = "0123456789abcdef") const
    {
        // print hex
        int k = 0;
        for (int i = 0; i < 5; i++) {
            for (int j = 7; j >= 0; j--) {
                hex[k++] = alphabet[(state[i] >> j * 4) & 0xf];
            }
        }
        if (zero_terminate)
            hex[k] = '\0';
        return *this;
    }

    const sha1& print_base64(char* base64, bool zero_terminate = true) const
    {
        static const uint8_t* table = (const uint8_t*)"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
                                                      "abcdefghijklmnopqrstuvwxyz"
                                                      "0123456789"
                                                      "+/";

        uint32_t triples[7] = {
            ((state[0] & 0xffffff00) >> 1 * 8),
            ((state[0] & 0x000000ff) << 2 * 8) | ((state[1] & 0xffff0000) >> 2 * 8),
            ((state[1] & 0x0000ffff) << 1 * 8) | ((state[2] & 0xff000000) >> 3 * 8),
            ((state[2] & 0x00ffffff) << 0 * 8),
            ((state[3] & 0xffffff00) >> 1 * 8),
            ((state[3] & 0x000000ff) << 2 * 8) | ((state[4] & 0xffff0000) >> 2 * 8),
            ((state[4] & 0x0000ffff) << 1 * 8),
        };

        for (int i = 0; i < 7; i++) {
            uint32_t x = triples[i];
            base64[i * 4 + 0] = table[(x >> 3 * 6) % 64];
            base64[i * 4 + 1] = table[(x >> 2 * 6) % 64];
            base64[i * 4 + 2] = table[(x >> 1 * 6) % 64];
            base64[i * 4 + 3] = table[(x >> 0 * 6) % 64];
        }

        base64[SHA1_BASE64_SIZE - 2] = '=';
        if (zero_terminate)
            base64[SHA1_BASE64_SIZE - 1] = '\0';
        return *this;
    }
};